File: server-go/internal/models/rtu_distribution.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type DistributionStatus string
type DistributionType string

const (
	DistributionStatusDraft     DistributionStatus = "DRAFT"
	DistributionStatusShipped   DistributionStatus = "SHIPPED"
	DistributionStatusReceived  DistributionStatus = "RECEIVED"
	DistributionStatusCancelled DistributionStatus = "CANCELLED"

	DistributionTypeProduct  DistributionType = "PRODUCT"
	DistributionTypeMaterial DistributionType = "MATERIAL"
)

type RTUDistribution struct {
	ID           uuid.UUID          `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"_id"`
	DocNumber    string             `gorm:"uniqueIndex;not null" json:"docNumber"`
	OutletID       *uuid.UUID         `gorm:"type:uuid" json:"outletId,omitempty"`
	Outlet         *Outlet            `gorm:"foreignKey:OutletID" json:"outlet,omitempty"`
	VendorID       *uuid.UUID         `gorm:"type:uuid;index" json:"vendorId,omitempty"`
	Vendor         *RTUVendor         `gorm:"foreignKey:VendorID" json:"vendor,omitempty"`
	SourceOutletID *uuid.UUID         `gorm:"type:uuid" json:"sourceOutletId,omitempty"`
	SourceOutlet   *Outlet            `gorm:"foreignKey:SourceOutletID" json:"sourceOutlet,omitempty"`
	Type           DistributionType   `gorm:"type:varchar(20);default:'PRODUCT'" json:"type"` // PRODUCT or MATERIAL
	Status       DistributionStatus `gorm:"type:string;default:'DRAFT'" json:"status"`
	ShipmentDate *time.Time         `json:"shipmentDate"`
	ReceivedDate *time.Time         `json:"receivedDate"`
	Notes        string             `json:"notes"`
	Items        []RTUDistributionItem `gorm:"foreignKey:DistributionID" json:"items"`

	IsDeleted bool       `gorm:"default:false" json:"isDeleted"`
	CreatedBy *uuid.UUID `gorm:"type:uuid" json:"createdBy,omitempty"`
	UpdatedBy *uuid.UUID `gorm:"type:uuid" json:"updatedBy,omitempty"`
	CreatedAt time.Time  `gorm:"autoCreateTime" json:"createdAt"`
	UpdatedAt time.Time  `gorm:"autoUpdateTime" json:"updatedAt"`

	Creator *User `gorm:"foreignKey:CreatedBy" json:"creator,omitempty"`
	Updater *User `gorm:"foreignKey:UpdatedBy" json:"updater,omitempty"`

	Histories []RTUDistributionHistory `gorm:"foreignKey:DistributionID" json:"histories"`
}

func (RTUDistribution) TableName() string { return "rtu_distributions" }

type RTUDistributionHistory struct {
	ID             uuid.UUID  `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"_id"`
	DistributionID uuid.UUID  `gorm:"type:uuid;not null;index" json:"distributionId"`
	Action         string     `gorm:"type:varchar(50);not null" json:"action"`
	Changes        string     `gorm:"type:text" json:"changes"`
	PerformedBy    *uuid.UUID `gorm:"type:uuid" json:"performedBy"`
	CreatedAt      time.Time  `gorm:"autoCreateTime" json:"createdAt"`

	Performer *User `gorm:"foreignKey:PerformedBy" json:"performer,omitempty"`
}

func (RTUDistributionHistory) TableName() string { return "rtu_distribution_histories" }

// RTUDistributionItem holds one line of goods in a shipment.
// Either ProductID or MaterialID is populated depending on the parent distribution's Type.
type RTUDistributionItem struct {
	ID             uuid.UUID    `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"_id"`
	DistributionID uuid.UUID    `gorm:"type:uuid;not null;index" json:"distributionId"`
	ProductID      *uuid.UUID   `gorm:"type:uuid;index" json:"productId,omitempty"`
	Product        *RTUProduct  `gorm:"foreignKey:ProductID" json:"product,omitempty"`
	MaterialID     *uuid.UUID   `gorm:"type:uuid;index" json:"materialId,omitempty"`
	Material       *RTUMaterial `gorm:"foreignKey:MaterialID" json:"material,omitempty"`
	Qty            float64      `gorm:"not null" json:"qty"`
	Unit           string       `json:"unit"`
	UnitPrice      float64      `gorm:"type:decimal(14,4);default:0" json:"unitPrice"` // FIFO cost at shipment time
	Notes          string       `json:"notes"`
}

func (RTUDistributionItem) TableName() string { return "rtu_distribution_items" }

File: client/lib/type/rtu_distribution.ts
import { RTUProduct } from './rtu_product';
import { RTUMaterial } from './rtu_material';
import { Outlet } from './outlet';
import { RTUVendor } from './rtu_vendor';

export type DistributionStatus = 'DRAFT' | 'SHIPPED' | 'RECEIVED' | 'CANCELLED';
export type DistributionType = 'PRODUCT' | 'MATERIAL';

export interface RTUDistributionItem {
  _id: string;
  distributionId: string;
  productId?: string;
  product?: RTUProduct;
  materialId?: string;
  material?: RTUMaterial;
  qty: number;
  unit: string;
  unitPrice?: number;
  notes?: string;
}

export interface RTUDistributionHistory {
  _id: string;
  distributionId: string;
  action: string;
  changes: string;
  performedBy?: string;
  createdAt: string;
  performer?: {
    _id: string;
    name: string;
  };
}

export interface RTUDistribution {
  _id: string;
  docNumber: string;
  outletId?: string;
  outlet?: Outlet;
  vendorId?: string;
  vendor?: RTUVendor;
  sourceOutletId?: string;
  sourceOutlet?: Outlet;
  type: DistributionType;
  status: DistributionStatus;
  shipmentDate?: string;
  receivedDate?: string;
  notes?: string;
  items: RTUDistributionItem[];
  isDeleted?: boolean;
  createdAt: string;
  updatedAt: string;
  creator?: {
    _id: string;
    name: string;
  };
  updater?: {
    _id: string;
    name: string;
  };
  histories?: RTUDistributionHistory[];
}
